cmd/cgr-tester/callproc.go

```diff
@@ -0,0 +1,165 @@
+
+	event := &utils.CGREvent{
+		Tenant: *tenant,
+		ID:     "TheEventID100000",
+		Time:   utils.TimePointer(time.Now()),
+		Event: map[string]any{
+			utils.AccountField: *subject,
+			utils.Destination:  *destination,
+			utils.OriginHost:   "local",
+			utils.RequestType:  utils.MetaRated,
+			utils.SetupTime:    st,
+			utils.Source:       "cgr_tester",
+			utils.OriginID:     utils.GenUUID(),
+		},
+		APIOpts: map[string]any{},
+	}
+
+	cpsCounter += 1
+	log.Printf("current call number: %+v", cpsCounter)
+
+	if *updateInterval > *maxUsage {
+		return fmt.Errorf(`"update_interval" should be smaller than "max_usage"`)
+	} else if *maxUsage <= *minUsage {
+		return fmt.Errorf(`"min_usage" should be smaller than "max_usage"`)
+	}
+
+	tstCfg.SessionSCfg().DebitInterval = 0
+
+	clntHandlers := map[string]any{utils.SessionSv1DisconnectSession: handleDisconnectSession}
+	brpc, err = utils.NewBiJSONrpcClient(tstCfg.SessionSCfg().ListenBijson, clntHandlers)
+	if err != nil {
+		return
+	}
+
+	//
+	// SessionSv1AuthorizeEvent
+	//
+	authArgs := &sessions.V1AuthorizeArgs{
+		GetMaxUsage: true,
+		CGREvent:    event,
+	}
+	var authRply sessions.V1AuthorizeReply
+	if err = brpc.Call(utils.SessionSv1AuthorizeEvent, authArgs, &authRply); err != nil {
+		return
+	}
+	// log.Printf("auth: %+v", utils.ToJSON(authRply))
+
+	//
+	// SessionSv1InitiateSession
+	//
+	event.Event[utils.AnswerTime] = at
+	event.Event[utils.RequestType] = utils.MetaRated
+
+	initArgs := &sessions.V1InitSessionArgs{
+		InitSession: true,
+		CGREvent:    event,
+	}
+
+	var initRply sessions.V1InitSessionReply
+	if err = brpc.Call(utils.SessionSv1InitiateSession, initArgs, &initRply); err != nil {
+		return
+	}
+	// log.Printf("init: %+v.", utils.ToJSON(initRply))
+
+	//
+	// SessionSv1UpdateSession
+	//
+	totalUsage := time.Duration(utils.RandomInteger(int64(*minUsage), int64(*maxUsage)))
+	log.Println("randUsage", totalUsage)
+	for currentUsage < totalUsage {
+		// log.Println("currentUsage", currentUsage)
+		currentUsage += *updateInterval
+		if currentUsage >= totalUsage {
+			break
+		}
+		event.Event[utils.Usage] = currentUsage.String()
+
+		upArgs := &sessions.V1UpdateSessionArgs{
+			GetAttributes: true,
+			UpdateSession: true,
+			CGREvent:      event,
+		}
+		var upRply sessions.V1UpdateSessionReply
+		if err = brpc.Call(utils.SessionSv1UpdateSession, upArgs, &upRply); err != nil {
+			return
+		}
+		// log.Printf("update: %+v.", utils.ToJSON(upRply))
+	}
+
+	//
+	// SessionSv1TerminateSession
+	//
+	event.Event[utils.Usage] = totalUsage.String()
+
+	tArgs := &sessions.V1TerminateSessionArgs{
+		TerminateSession: true,
+		CGREvent:         event,
+	}
+	var tRply string
+	if err = brpc.Call(utils.SessionSv1TerminateSession, tArgs, &tRply); err != nil {
+		return
+	}
+	// log.Printf("terminate: %+v.", utils.ToJSON(tRply))
+
+	//
+	// SessionSv1ProcessCDR
+	//
+	procArgs := event
+	var pRply string
+	if err = brpc.Call(utils.SessionSv1ProcessCDR, procArgs, &pRply); err != nil {
```

Comment:
Please add a small random delay between terminate and processCDR to get closer to reality.